text/1021-rewrites.rst

```diff
@@ -0,0 +1,210 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2023-01-12
+    Authors: Michel J. Sullivan <sully@msully.net>
+
+===========================
+RFC 1021: Mutation rewrites
+===========================
+
+This RFC proposes a mechanism for specifying rewrite rules that are
+applied to properties and links on INSERT and UPDATE of objects.
+
+They can be thought of as a generalization of ``default``.
+
+Motivation
+==========
+
+It is sometimes useful to have a pointer in an object automatically updated
+whenever the object has an ``UPDATE`` performed on it.
+
+The most straightforward example of this is having a
+``modification_time`` field that tracks when an object was last
+modified.
+
+It is also valuable when (probably for performance reasons) you wish
+to maintain a field that is derived from other object
+state. (For example, maintaining a count of friends.)
+
+Note
+====
+
+This is a companion to
+`RFC 1020 <https://github.com/edgedb/rfcs/blob/master/text/1020-triggers.rst>`_,
+in that they work together to cover most of the desired use cases that motivated
+a previous `previous version triggers RFC <https://github.com/edgedb/rfcs/pull/70>`_
+
+Specification
+=============
+
+Synopsis::
+
+    type <type-name> "{"
+        { property | link } <name> -> <type> "{"
+            rewrite { insert | update } [, ... ] using (<expr>)
+        "}"
+    "}"
+
+We allow pointer definitions to be augmented with a ``rewrite`` rule.
+
+Expression interpretation
+-------------------------
+In ``<expr>``, ``__subject__`` (and the implicit path prefix) will be
+bound to the new value object being inserted or updated.
+
+In purely ``update`` rewrite rules, ``__old__`` will refer to the old
+version of the object.
+
+The special variable ``__fields__`` will be bound to all of the
+properties of ``<type-name>`` that were explicitly provided in the
+``insert`` or ``update``. This allows rewrite rules to distinguish
+between an explicitly provided empty set and no value being specified.
```

Comment:
Will `__fields__` contain values of other pointers that have already been rewritten?

A use-case:
```
type TodoList {
    multi link items -> Items;
    property is_empty -> bool {
        rewrite insert, update using (count(__subject__.items) = 0);
    }
    property status -> str {
        rewrite insert, update using (
            'done' if __subject__.is_empty else __subject__.status
        );
    }
}
```

If yes, in what order will they be evaluated and can they contain cycles?

The answer is obvious, the question is: do we want to support that?

Reply:
Hmmm. My intention was that all of the rewrites would be applied simultaneously, and all of them would refer to the original object

Reply:
can we call it `__values__`?

Reply:
Another way would be to have __specified__ be a named tuple that contains booleans for each of the pointers:
```
... using ('had_status' if __specified__.status else 'did_not_have_status')
```

Reply:
```
        rewrite on insert on update using (
            'done' if __subject__.is_empty else __subject__.status
        );
```

Reply:
```
        rewrite (insert, update) using (
            'done' if __subject__.is_empty else __subject__.status
        );
```

Reply:
```
        rewrite insert and update using (
            'done' if __subject__.is_empty else __subject__.status
        );
```

Reply:
```
        rewrite insert update using (
            'done' if __subject__.is_empty else __subject__.status
        );
```